dev/tools/pre-commit

```diff
@@ -7,10 +7,25 @@ set -e
 
 dev/tools/check-overlays.sh
 
+# Can we check and fix formatting?
+ocamlformat=$(command -v ocamlformat || echo true)
+if [ "$ocamlformat" = true ]
+then
+    1>&2 echo "Warning: ocamlformat is not in path. Cannot check formatting."
+fi
+
+# Is the syntax correct?
+if ! git diff --cached --name-only -z | grep -E '.*\.mli?$' -z | xargs -0 $ocamlformat
+then
+    ocamlformat=true
+    1>&2 echo "Warning: some files are not syntactically valid. Formatting will not be checked."
+fi
+
 if ! git diff --cached --name-only -z | xargs -0 dev/tools/check-eof-newline.sh ||
+        ! git diff --cached --name-only -z | grep -E '.*\.mli?$' -z | xargs -0 $ocamlformat --check ||
         ! git diff-index --check --cached HEAD >/dev/null 2>&1 ;
```

Comment:
>

Actually @SkySkimmer the problem you describe in your scenario exists even without this preliminary check. Suppose I git add a file with broken syntax, then fix it in the worktree (including formatting), then git commit without option -a. The pre-commit hook will skip the fixing step and thus my commit will have bad formatting.

True, I guess we should just remove these tests and run unconditionally.

Reply:
This PR is taking me far beyond what I was expecting. I was just hoping to help by providing a quick patch over the weekend. Would you like to finish it?

Reply:
OK